plugins/ltac/rewrite.ml

```diff
@@ -417,11 +427,18 @@ module TypeGlobal = struct
 
 
   let inverse env (evd,cstrs) car rel =
+    let evd, car = Evarsolve.refresh_universes ~onlyalg:true (Some false) env evd car in
     let (evd, sort) = Evarutil.new_Type ~rigid:Evd.univ_flexible evd in
       app_poly_check env (evd,cstrs) coq_inverse [| car ; car; sort; rel |]
 
 end
 
+let get_type_of_refresh env evars t =
```

Comment:
There's already a function of this name in evarsolve which should be used (it needs to be extended a bit to understand onlyalg and status)

Reply:
I've made this a wrapper around `Evarsolve.get_type_of_refresh` the fact that it is passing around the snd projection but doing nothing is a bit tricky to factor out in later code. It is very confusing to deal with `evars` when it can mean `Evd.evar_map * Evar.Set.t` or `Evd.evar_map`. Any suggestions on how to best refactor the uses?

Reply:
I think it is fine as is. We need to keep this set of evars correponding to constraints along the evar_map so we do not accidentally try to resolve them in the first phase of the rewrite when we might launch typeclass resolution after unification of the rewrite lemma.